service/dap/server.go

```diff
@@ -442,9 +455,135 @@ func (s *Server) onLaunchRequest(request *dap.LaunchRequest) {
 		return
 	}
 
-	mode, ok := request.Arguments["mode"]
-	if !ok || mode == "" {
-		mode = "debug"
+	if mode == "record" {
+		backend, ok := request.Arguments["backend"].(string)
+		if !ok || backend == "" {
+			s.sendErrorResponse(request.Request,
+				FailedToLaunch, "Failed to launch",
+				"The backend attribute is missing debug configuration.")
+			return
+		}
+
+		// Validate backend for mode
+		if backend != "rr" {
+			s.sendErrorResponse(request.Request,
+				FailedToLaunch, "Failed to launch",
+				fmt.Sprintf("Unsupported 'backend' value %q in debug configuration.", backend))
+			return
+		}
+
+		// Record only works on rr trace directories
+		traceDirectory, ok := request.Arguments["traceDirectory"].(string)
+		if !ok || traceDirectory == "" {
+			s.sendErrorResponse(request.Request,
+				FailedToLaunch, "Failed to launch",
+				"The traceDirectory attribute is missing in debug configuration.")
+			return
+		}
+		if _, err := os.Stat(traceDirectory); os.IsNotExist(err) {
+			s.sendErrorResponse(request.Request,
+				FailedToLaunch, "Failed to launch",
+				"The traceDirectory attribute points to an directory that does not exist.")
+			return
+		}
+
+		buildFlags := ""
+		buildFlagsArg, ok := request.Arguments["buildFlags"]
+		if ok {
+			buildFlags, ok = buildFlagsArg.(string)
+			if !ok {
+				s.sendErrorResponse(request.Request,
+					FailedToLaunch, "Failed to launch",
+					fmt.Sprintf("'buildFlags' attribute '%v' in debug configuration is not a string.", buildFlagsArg))
+				return
+			}
+		}
+
+		output, ok := request.Arguments["output"].(string)
+		if !ok || output == "" {
+			output = debugBinary
+		}
+		debugname, err := filepath.Abs(output)
+		err = gobuild.GoBuild(debugname, []string{program}, buildFlags)
+
+		if err != nil {
+			s.sendErrorResponse(request.Request,
+				FailedToLaunch, "Failed to build recording target",
+				fmt.Sprintf("Build error: %s", err.Error()))
+			return
+		}
+	}
+
+	if mode == "coredump" {
+		backend, ok := request.Arguments["backend"].(string)
+		if !ok || backend == "" {
+			s.sendErrorResponse(request.Request,
+				FailedToLaunch, "Failed to launch",
+				"The backend attribute is missing debug configuration.")
+			return
+		}
+
+		// Validate backend for mode
+		if backend != "coredump" {
```

Comment:
This was corrected, it was a wrong assumption of mine from the old pr. I assume 'core' as the alternative backend to rr for the case of core dumps because, even if it is not registered or documented as a native backend per se, that's how delve handles it when asked about it:

https://github.com/go-delve/delve/blob/fe904c14d1c4984ca52b66cc8626362597094d21/service/debugger/debugger.go#L1610-L1616

We can omit the backend for core dumps if you want, but I think it may be better in the long run to make this explicit for the users and having 2 distinct backends for their use cases

Reply:
I agree that the discrepancy is unfortunate. Should we be making the same argument for dlv as well then? Ideally we want the clients and the interfaces to be aligned, so the users can easily apply their knowledge of one to the other. There is no button in vscode for the users to ask for the version, so they never see this discrepancy, will they? But they will notice a difference in the launch args. 

The "core" value is communicated via the request, but is not used internally as far as I can tell.
For example this switch would treat "core" or "coredump" that you introduced as an unknown backend.
https://github.com/go-delve/delve/blob/333e84a0cb209a9ec762b753cbd4ecbc679104d4/service/debugger/debugger.go#L245-L295

Reply:
Yeah, the core `backend` parameter from the vs code request serves just as a way to call `dlv core`, which itself sets up a debugger `CoreFile` and `Backend` parameters internally. To a vs code only user may look simpler but to an advanced user it will cause more confusion, I concede that.

So, for practical matters, I could manage with just the `coreDumpPath` parameter I receive from vs code on `replay` mode and with that infer that we are going to replay a core dump if it is present on the request and it is not empty, otherwise we can assume it is going to be an rr trace